src/devices/Gpio/Drivers/Sunxi/SunxiDriver.cs

```diff
@@ -0,0 +1,468 @@
+    /// It can even drive the internal pins that are not drawn out.
+    /// Before you operate, you must be clear about what you are doing.
+    /// </remarks>
+    public unsafe class SunxiDriver : SysFsDriver
+    {
+        private const string GpioMemoryFilePath = "/dev/mem";
+        private readonly IDictionary<int, PinState> _pinModes = new Dictionary<int, PinState>();
+
+        private IntPtr _gpioPointer0;
+        private IntPtr _gpioPointer1;
+
+        /// <summary>
+        /// CPUX-PORT base address.
+        /// </summary>
+        protected virtual int CpuxPortBaseAddress { get; }
+
+        /// <summary>
+        /// CPUS-PORT base address.
+        /// </summary>
+        protected virtual int CpusPortBaseAddress { get; }
+
+        // final_address = mapped_address + (target_address & map_mask) https://stackoverflow.com/a/37922968
+        private readonly int _mapMask = Environment.SystemPageSize - 1;
+        private static readonly object s_initializationLock = new object();
+        private static readonly object s_sysFsInitializationLock = new object();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SunxiDriver"/> class.
+        /// </summary>
+        protected SunxiDriver()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SunxiDriver"/>.
+        /// </summary>
+        /// <param name="cpuxPortBaseAddress">CPUX-PORT base address (This can be find in the corresponding SoC datasheet).</param>
+        /// <param name="cpusPortBaseAddress">CPUS-PORT base address (This can be find in the corresponding SoC datasheet).</param>
+        public SunxiDriver(int cpuxPortBaseAddress, int cpusPortBaseAddress)
+        {
+            CpuxPortBaseAddress = cpuxPortBaseAddress;
+            CpusPortBaseAddress = cpusPortBaseAddress;
+        }
+
+        /// <summary>
+        /// The number of pins provided by the driver.
+        /// </summary>
+        protected override int PinCount => throw new PlatformNotSupportedException("This driver is generic so it can not enumerate how many pins are available.");
+
+        /// <summary>
+        /// Converts a board pin number to the driver's logical numbering scheme.
+        /// </summary>
+        /// <param name="pinNumber">The board pin number to convert.</param>
+        /// <returns>The pin number in the driver's logical numbering scheme.</returns>
+        protected override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber) => throw new PlatformNotSupportedException("This driver is generic so it can not perform conversions between pin numbering schemes.");
+
+        /// <summary>
+        /// Opens a pin in order for it to be ready to use.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the driver's logical numbering scheme.</param>
+        protected override void OpenPin(int pinNumber)
+        {
+            Initialize();
```

Comment:
I think initialize should be called from the constructor instead because you can't prevent from folks calling ClosePin before calling OpenPin. Also seems like any operation will require Initialize to have happened, so no need to delay it to the first OpenPin call.